mono/metadata/sgen-mono.c

```diff
@@ -2217,14 +2217,127 @@ sgen_client_scan_thread_data (void *start_nursery, void *end_nursery, gboolean p
+		// void f()
+		// {
+		// 	int local [5];
+		// 
+		// 	while (unlucky && ((size_t)_AddressOfReturnAddress () - 8) & 0xFFF)
+		// 		f ();
+		// 
+		// 	unlucky = 0;
+		// 	Event1 = local;
+		// 
+		// 	while (!Event2)
+		// 		_mm_pause ();
+		// 
+		// 	if (print) {
+		// 		printf ("%X\n", local [0]);
+		// 		print = 0;
+		// 	}
+		// 
+		// 	if (ThreadHandle) {
+		// 		WaitForSingleObject (ThreadHandle, INFINITE);
+		// 		ThreadHandle = NULL;
+		// 	}
+		// }
+		// 
+		// int main (int argc, char** argv)
+		// {
+		// 	unlucky = argc > 1;
+		// 	ThreadHandle = CreateThread (0, 0, thread, 0, 0, 0);
+		// 	f ();
+		// }
+		//
+		// This would seem to be a problem otherwise, not just for garbage collectors.
+		//
+		// However we can do better.
+		// The extent that the stack has been commited, that guard pages
+		// have been touched and moved down, is recorded in NtCurrentTeb ()->StackLimit.
+		//
+		// So rather than pay for the VirtualQuery syscall, we can instead
+		// use the max of rsp and StackLimit. Nothing below StackLimit should be touched
+		// by the garbage collector. Nothing below StackLimit has been touched by the suspended thread.
+		//
+		// The inaccuracy in the comment is that, a crash is easily avoided.
+		// This thread can catch the guard page exception and reset the guard page.
+		//	Race condition, except, the other thread is suspended.
+		// The problem with that is that __try / __except does not work with gcc. It is also slow,
+		// but exceedingly rarely so.
+		//
+		// There is a narrow window generally between stack adjustments
+		// and local variable access, and this is only a problem when reaching new high water marks,
+		// and it only matters when return address is on a different page than locals. The combination
+		// of all three factors is presumably rare but cannot be ruled out.
+		//
+		// A vectored exception handler could also be used.
+		//
+		// VirtualProtect was not allowed in UWP but will be. There is also VirtualProtectFromApp.
+#if 0
 		MEMORY_BASIC_INFORMATION mem_info;
-		SIZE_T result = VirtualQuery(info->client_info.stack_start, &mem_info, sizeof(mem_info));
+		const SIZE_T result = VirtualQuery(info->client_info.stack_start, &mem_info, sizeof(mem_info));
 		g_assert (result != 0);
 		if (mem_info.Protect & PAGE_GUARD) {
 			aligned_stack_start = ((char*) mem_info.BaseAddress) + mem_info.RegionSize;
 		}
+#else // old code vs. new code
+#if 0 // _MSC_VER
+		__try {
+			*(volatile char*)aligned_stack_start;
+		} __except (GetExceptionCode () == STATUS_GUARD_PAGE_VIOLATION) {
+			MEMORY_BASIC_INFORMATION mem_info;
+			const SIZE_T result = VirtualQuery(aligned_stack_start, &mem_info, sizeof(mem_info));
+			g_assert (result >= sizeof (mem_info));
+			VirtualProtect (aligned_stack_start, 1, mem_info.Protect | PAGE_GUARD, &mem_info.Protect);
+		}
+#else // clang / gcc
+		aligned_stack_start = MAX (aligned_stack_start, ((PNT_TIB)info->client_info.info.windows_teb)->StackLimit);
```

Comment:
I thought Windows tries to discourage people from accessing members in a TEB directly? Does that not apply here because you just care about the TIB? I'm far from an expert on Windows, so I'd be curious what the specifics are for that.

Reply:
I pretty sure TIB is public, TEB is private. Winnt.h vs. winternl.h. I can try to check.

Reply:
I pretty sure TIB is public, TEB is private. Winnt.h vs. winternl.h. I can try to check.

Reply:
I pretty sure TIB is public, TEB is private. Winnt.h vs. winternl.h. I can try to check.

Reply:
I confirmed it is ok.